cvmfs/fetch.cc

```diff
@@ -94,31 +98,32 @@ int Fetcher::Fetch(
 
   ThreadLocalStorage *tls = GetTls();
 
-  // Synchronization point: either act as a master thread for this object or
-  // enqueue to the list of waiting threads.
-  pthread_mutex_lock(lock_queues_download_);
-  ThreadQueues::iterator iDownloadQueue = queues_download_.find(id);
-  if (iDownloadQueue != queues_download_.end()) {
-    LogCvmfs(kLogCache, kLogDebug, "waiting for download of %s", name.c_str());
-
-    iDownloadQueue->second->push_back(tls->pipe_wait[1]);
-    pthread_mutex_unlock(lock_queues_download_);
-    ReadPipe(tls->pipe_wait[0], &fd_return, sizeof(int));
-
-    LogCvmfs(kLogCache, kLogDebug, "received from another thread fd %d for %s",
-             fd_return, name.c_str());
-    return fd_return;
-  } else {
-    // Seems we are the first one, check again in the cache (race condition)
-    fd_return = OpenSelect(id, name, object_type);
-    if (fd_return >= 0) {
-      pthread_mutex_unlock(lock_queues_download_);
+  {
+    // Synchronization point: either act as a master thread for this object or
+    // enqueue to the list of waiting threads.
+    MutexLockGuard m(lock_queues_download_);
+    ThreadQueues::iterator iDownloadQueue = queues_download_.find(id);
+    if (iDownloadQueue != queues_download_.end()) {
+      LogCvmfs(kLogCache, kLogDebug, "waiting for download of %s",
+               name.c_str());
+
+      iDownloadQueue->second->push_back(tls->pipe_wait[1]);
+      ReadPipe(tls->pipe_wait[0], &fd_return, sizeof(int));
```

Comment:
Also here, the call to ReadPipe is outside the lock in the original code while I put it inside.

I don't believe is a problem but is worth a second look...